clients/pkg/promtail/client/watcher.go

```diff
@@ -0,0 +1,333 @@
+// for log WAL entries, but also, the consumer surface has been implemented according to the actions necessary for
+// Promtail's WAL.
+
+// WALReader is a dependency interface to inject generic WAL readers into the WALWatcher.
+type WALReader interface {
+	Next() bool
+	Err() error
+	Record() []byte
+}
+
+// WatcherConsumer is responsible for doing the necessary work to process both series and entries while the WALWatcher
+// is reading / tailing segments. Also, when a new segment is detected, and the watcher is moving on from one to the other,
+// the SegmentEnd callback is available to take action if required.
+// The implementor of this interface is not expected to implement thread safety if used on a single watcher, since each
+// callback is called synchronously.
+type WatcherConsumer interface {
+	ConsumeSeries(series record.RefSeries) error
+	ConsumeEntries(entries ingester.RefEntries) error
+	SegmentEnd(segmentNum int)
+}
+
+type WALWatcher struct {
+	name       string
+	consumer   WatcherConsumer
+	done       chan struct{}
+	quit       chan struct{}
+	walDir     string
+	logger     log.Logger
+	MaxSegment int
+
+	metrics *WALWatcherMetrics
+}
+
+func NewWALWatcher(walDir, name string, metrics *WALWatcherMetrics, consumer WatcherConsumer, logger log.Logger) *WALWatcher {
+	return &WALWatcher{
+		walDir:     walDir,
+		name:       name,
+		consumer:   consumer,
+		quit:       make(chan struct{}),
+		done:       make(chan struct{}),
+		MaxSegment: -1,
+		logger:     logger,
+		metrics:    metrics,
+	}
+}
+
+// Start runs the watcher main loop.
+func (w *WALWatcher) Start() {
+	w.metrics.watchersRunning.WithLabelValues().Inc()
+	go w.mainLoop()
+}
+
+// mainLoop retries when there's an error reading a specific segment or advancing one, but leaving a bigger time in-between
+// retries.
+func (w *WALWatcher) mainLoop() {
+	defer close(w.done)
+	for !isClosed(w.quit) {
+		if err := w.run(); err != nil {
+			level.Error(w.logger).Log("msg", "error tailing WAL", "err", err)
+		}
+
+		select {
+		case <-w.quit:
+			return
+		case <-time.After(5 * time.Second):
+		}
+	}
+}
+
+// Run the watcher, which will tail the WAL until the quit channel is closed
+// or an error case is hit.
+func (w *WALWatcher) run() error {
+	_, lastSegment, err := w.firstAndLast()
+	if err != nil {
+		return fmt.Errorf("wal.Segments: %w", err)
+	}
+
+	currentSegment := lastSegment
+	level.Debug(w.logger).Log("msg", "Tailing WAL", "currentSegment", currentSegment, "lastSegment", lastSegment)
+	for !isClosed(w.quit) {
+		w.metrics.currentSegment.WithLabelValues(w.name).Set(float64(currentSegment))
+		level.Debug(w.logger).Log("msg", "Processing segment", "currentSegment", currentSegment)
+
+		// On start, we have a pointer to what is the latest segment. On subsequent calls to this function,
+		// currentSegment will have been incremented, and we should open that segment.
+		if err := w.watch(currentSegment, currentSegment >= lastSegment); err != nil {
```

Comment:
```suggestion
		if err := w.watch(currentSegment, true); err != nil {
```
Isn't `currentSegment >= lastSegment` always going to be true or am I misreading the code here? `currentSegment` starts out equal to `lastSegment` and only ever gets larger while lastSegment seems to stay the same.

Reply:
Uhh you are right! That bit comes from Prometheus'es WAL watcher in which I based this implementation. In their case, the watcher itself replays the WAL, but in our case, we just watch from the last segment onwards. Cleaning up here all logic related to replaying from the watcher itself. Should simplify a little bit.

Reply:
I'm still a bit confused here as well. We need to replay the WAL in case anything is unsent and also to cache labels data for streams that already existed prior to the most recent start up. I'm not seeing how we do that caching and reuse of stream ref IDs.